builtin/logical/aws/client.go

```diff
@@ -24,19 +24,79 @@ import (
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
-const fallbackEndpoint = "https://sts.amazonaws.com" // this is not regionally distributed; all requests go to us-east-1
+// getRootIAMConfig creates an *aws.Config for Vault to connect to IAM.
+func (b *backend) getRootIAMConfig(ctx context.Context, s logical.Storage, logger hclog.Logger) (*aws.Config, error) {
+	credsConfig := &awsutil.CredentialsConfig{}
+	var endpoint string
+	var maxRetries int = aws.UseServiceDefaultRetries
+
+	entry, err := s.Get(ctx, "config/root")
+	if err != nil {
+		return nil, err
+	}
+	if entry != nil {
+		var config rootConfig
+		if err := entry.DecodeJSON(&config); err != nil {
+			return nil, fmt.Errorf("error reading root configuration: %w", err)
+		}
+
+		credsConfig.AccessKey = config.AccessKey
+		credsConfig.SecretKey = config.SecretKey
+		credsConfig.Region = config.Region
+		maxRetries = config.MaxRetries
+
+		if config.IAMEndpoint != "" {
+			endpoint = *aws.String(config.IAMEndpoint)
+		}
+
+		if config.IdentityTokenAudience != "" {
+			ns, err := namespace.FromContext(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get namespace from context: %w", err)
+			}
+
+			fetcher := &PluginIdentityTokenFetcher{
+				sys:      b.System(),
+				logger:   b.Logger(),
+				ns:       ns,
+				audience: config.IdentityTokenAudience,
+				ttl:      config.IdentityTokenTTL,
+			}
+
+			sessionSuffix := strconv.FormatInt(time.Now().UnixNano(), 10)
+			credsConfig.RoleSessionName = fmt.Sprintf("vault-aws-secrets-%s", sessionSuffix)
+			credsConfig.WebIdentityTokenFetcher = fetcher
+			credsConfig.RoleARN = config.RoleARN
+		}
+	}
+
+	if credsConfig.Region == "" {
+		credsConfig.Region = getFallbackRegion()
+	}
+
+	credsConfig.HTTPClient = cleanhttp.DefaultClient()
+
+	credsConfig.Logger = logger
+
+	creds, err := credsConfig.GenerateCredentialChain()
+	if err != nil {
+		return nil, err
+	}
+
+	return &aws.Config{
+		Credentials: creds,
+		Region:      aws.String(credsConfig.Region),
+		Endpoint:    &endpoint,
+		HTTPClient:  cleanhttp.DefaultClient(),
+		MaxRetries:  aws.Int(maxRetries),
+	}, nil
+}
 
 // Return a slice of *aws.Config, based on descending configuration priority. STS endpoints are the only place this is used.
 // NOTE: The caller is required to ensure that b.clientMutex is at least read locked
-func (b *backend) getRootConfigs(ctx context.Context, s logical.Storage, clientType string, logger hclog.Logger) ([]*aws.Config, error) {
+func (b *backend) getRootSTSConfigs(ctx context.Context, s logical.Storage, logger hclog.Logger) ([]*aws.Config, error) {
```

Comment:
Nit: Should this be renamed to be singular instead of plural?

Reply:
this one does still return an array, so it returns multiple configs.

Reply:
@kpcraig um actually it's a slice

Reply:
Oh yeah I didn't catch that, sounds good.